File: src/runtime/boxes/BbSeq.cs
// BbSeq.cs — SEQ: concatenation — left then right; β retries right then left
// Mirrors src/runtime/boxes/bb_seq.c
//
// State machine (mirrors the C goto model exactly):
//   α:        try left.α
//             left_γ → try right.α
//             right_γ → SEQ_γ
//             right_ω → left.β → left_γ / left_ω
//             left_ω  → SEQ_ω
//   β:        try right.β
//             right_γ → SEQ_γ
//             right_ω → left.β → ...

namespace Snobol4.Runtime.Boxes;

public sealed class BbSeq : IByrdBox
{
    private readonly IByrdBox _left;
    private readonly IByrdBox _right;
    private Spec _matched;

    public BbSeq(IByrdBox left, IByrdBox right)
    {
        _left  = left;
        _right = right;
    }

    public Spec Alpha(MatchState ms)
    {
        _matched = Spec.ZeroWidth(ms.Cursor);
        return TryLeft(ms, fromAlpha: true);
    }

    public Spec Beta(MatchState ms)
    {
        return TryRight(ms, fromAlpha: false);
    }

    // ── internal state machine ──────────────────────────────────────────

    private Spec TryLeft(MatchState ms, bool fromAlpha)
    {
        var lr = fromAlpha ? _left.Alpha(ms) : _left.Beta(ms);
        if (lr.IsFail) return Spec.Fail;                    // left_ω → SEQ_ω
        // left_γ
        _matched = _matched.Cat(lr);
        return TryRight(ms, fromAlpha: true);
    }

    private Spec TryRight(MatchState ms, bool fromAlpha)
    {
        while (true)
        {
            var rr = fromAlpha ? _right.Alpha(ms) : _right.Beta(ms);
            if (!rr.IsFail)
                return _matched.Cat(rr);                    // right_γ → SEQ_γ

            // right_ω → retry left with β
            var lr = _left.Beta(ms);
            if (lr.IsFail) return Spec.Fail;                // left_ω → SEQ_ω
            _matched = _matched.Cat(lr);
            fromAlpha = true;                               // left_γ → retry right.α
        }
    }
}

File: src/runtime/boxes/BbSeq.java
/**
 * BbSeq.java — SEQ: concatenation; left then right; β retries right then left
 * Port of bb_seq.c / bb_seq.s
 *
 * C original:
 *   SEQ_α:    matched=spec(Σ+Δ,0);
 *             lr = left(α);   if empty → left_ω;  else → left_γ;
 *   SEQ_β:    rr = right(β);  if empty → right_ω; else → right_γ;
 *   left_γ:   matched = cat(matched, lr);
 *             rr = right(α);  if empty → right_ω; else → right_γ;
 *   left_ω:                                        goto SEQ_ω;
 *   right_γ:  SEQ = cat(matched, rr);              goto SEQ_γ;
 *   right_ω:  lr = left(β);   if empty → left_ω;  else → left_γ;
 *   SEQ_γ:                                         return SEQ;
 *   SEQ_ω:                                         return spec_empty;
 */
class BbSeq extends BbBox {
    private final BbBox left, right;
    private int matchedStart, matchedLen;  /* accumulated span */

    public BbSeq(MatchState ms, BbBox left, BbBox right) {
        super(ms);
        this.left  = left;
        this.right = right;
    }

    @Override public Spec alpha() {
        // SEQ_α
        matchedStart = ms.delta;
        matchedLen   = 0;
        Spec lr = left.alpha();
        if (lr == null) return null;                                   // left_ω → SEQ_ω
        // left_γ
        matchedLen += lr.len;
        return rightAlpha();
    }

    @Override public Spec beta() {
        // SEQ_β: retry right first
        Spec rr = right.beta();
        if (rr != null) return new Spec(matchedStart, matchedLen + rr.len); // right_γ
        // right_ω: backtrack left
        return leftBeta();
    }

    /* right_ω path: try left.β then right.α */
    private Spec leftBeta() {
        Spec lr = left.beta();
        if (lr == null) return null;                                   // left_ω → SEQ_ω
        // left_γ
        matchedLen = lr.len;
        return rightAlpha();
    }

    private Spec rightAlpha() {
        Spec rr = right.alpha();
        if (rr == null) return leftBeta();                             // right_ω
        return new Spec(matchedStart, matchedLen + rr.len);           // right_γ → SEQ_γ
    }
}
